### DISCUSSION TITLE
Support of non-scalar subselects

### DISCUSSION THREAD

**Discussion by david542542** [duckdb/duckdb#3568] [General]
Postgres, Snowflake and a few other DBs support non-scalar subselects, such as:

```
SELECT * FROM (VALUES (1,2.5,'this is a test'), (2,3,'asdf')) AS t(a,b,c)
WHERE 
    (a,b)=(1,2.5) AND 
    (a,c) in (select 1, 'this is a test') -- this part errors
```

However in DuckDB it gives a Binder error: 

> Binder Error: Subquery returns 2 columns - expected 1
 LINE 1: ...t(a,b,c) where (a,b)=(1,2.5) and (a,c) in (select 1, 'this is a test');

Are there any plans to support this feature? What do you estimate it might take to implement this?

**Comment by hannes** [comment]
I would say it's not an immediate priority, and you can of course express this as a join already. What is your use case you need this for?

**Comment by david542542** [comment]
Hi @hannes thanks for the response. To be honest, I've never used these in practice and I would probably just re-write it as two scalar subselects, i.e.,

```
-- ...(a,c) in (select 1, 'this is a test') -- this part errors
...a in (select 1) and c in (select 'test is a test')
```

I was more curious from a parsing perspective. I was working on a parser for DuckDB's `SELECT, and using Postgres where I can't find exact examples from DuckDB, just checking if that feature is available in DuckDB or not. Speaking of which, I think it'd be great to have a more formal lexical syntax in the docs, something like: https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical or https://www.postgresql.org/docs/current/sql-syntax.html. Usually when I can't find something I use a combination of Postgres docs and trial-and-error, but I think adding the formal part to the docs would be awesome and help a lot of people.

**Comment by orzom411** [comment]
Workaround, try this:

SELECT * 
  FROM (VALUES (1,2.5,'this is a test'), (2,3,'asdf')) AS t(a,b,c)
 WHERE 
    (a,b)=(1,2.5) 
    AND EXISTS (
     select 1 as e, 'this is a test' as f 
      where (e,f) = (a,c)
      ) -- Mitigate via exists island.
;

results:
a|b|c
1|2.5|this is a test

Little long hand, aliases matter or in this case I simply employed unique field names.  I like the way DuckDB handles, I'll call it shaped data.
